src/pe/certificate_table.rs

```diff
@@ -0,0 +1,90 @@
+
+#[repr(u16)]
+#[derive(Debug, PartialEq, Copy, Clone)]
+enum CertificateAttributeRevision {
+    Revision1_0 = 0x0100,
+    Revision2_0 = 0x0200,
+}
+
+impl TryFrom<u16> for CertificateAttributeRevision {
+    type Error = error::Error;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        Ok(match value {
+            x if x == CertificateAttributeRevision::Revision1_0 as u16 => CertificateAttributeRevision::Revision1_0,
+            x if x == CertificateAttributeRevision::Revision2_0 as u16 => CertificateAttributeRevision::Revision2_0,
+            _ => return Err(error::Error::Malformed("Invalid certificate attribute revision".to_string()))
+        })
+    }
+}
+
+#[repr(u16)]
+#[derive(Debug)]
+enum CertificateAttributeType {
+    X509 = 0x0001,
+    PkcsSignedData = 0x0002,
+    Reserved1 = 0x0003,
+    TsStackSigned = 0x0004,
+}
+
+impl TryFrom<u16> for CertificateAttributeType {
+    type Error = error::Error;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        Ok(match value {
+            x if x == CertificateAttributeType::X509 as u16 => CertificateAttributeType::X509,
+            x if x == CertificateAttributeType::PkcsSignedData as u16 => CertificateAttributeType::PkcsSignedData,
+            x if x == CertificateAttributeType::Reserved1 as u16 => CertificateAttributeType::Reserved1,
+            x if x == CertificateAttributeType::TsStackSigned as u16 => CertificateAttributeType::TsStackSigned,
+            _ => return Err(error::Error::Malformed("Invalid certificate attribute type".to_string()))
+        })
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, Pread)]
+pub struct CertificateAttributeHeader {
+    length: u32,
+    revision: u16,
+    certificate_type: u16
+}
+
+pub const CERTIFICATE_DATA_OFFSET: u32 = 8;
+#[derive(Debug)]
+pub struct CertificateAttribute<'a> {
+    length: u32,
+    revision: CertificateAttributeRevision,
+    certificate_type: CertificateAttributeType,
+    certificate: &'a [u8]
+}
+
+impl<'a> CertificateAttribute<'a> {
+    pub fn parse(bytes: &'a [u8], current_offset: &mut usize) -> Result<CertificateAttribute<'a>, error::Error> {
+        let header: CertificateAttributeHeader = bytes.gread_with(current_offset, scroll::LE)?;
+        let mut cert_size = (header.length - CERTIFICATE_DATA_OFFSET) as usize;
+        // Round to the next 8-bytes.
+        cert_size = (cert_size + 7) & !7;
+        let attr = Self {
+            length: header.length,
+            revision: header.revision.try_into()?,
+            certificate_type: header.certificate_type.try_into()?,
+            certificate: &bytes[*current_offset..(*current_offset + cert_size)]
+        };
+        *current_offset += cert_size as usize;
+        Ok(attr)
+    }
+}
+
+pub type CertificateDirectoryTable<'a> = Vec<CertificateAttribute<'a>>;
+pub fn enumerate_certificates<'a>(bytes: &'a [u8], table_virtual_address: u32, table_size: u32) -> Result<CertificateDirectoryTable<'a>, error::Error> {
```

Comment:
nit: you should be able to omit the `'a` annotation

Reply:
Addressed